lib/string/strchr/stpcspn.h

```diff
@@ -0,0 +1,25 @@
+// SPDX-FileCopyrightText: 2024, Alejandro Colomar <alx@kernel.org>
+// SPDX-License-Identifier: BSD-3-Clause
+
+
+#ifndef SHADOW_INCLUDE_LIB_STRING_STRCHR_STPCSPN_H_
+#define SHADOW_INCLUDE_LIB_STRING_STRCHR_STPCSPN_H_
+
+
+#include <config.h>
+
+#include <string.h>
+
+#include "attr.h"
+
+
+// Similar to strcspn(3), but return a pointer.
```

Comment:
... to the end?

Reply:
Yes.

Although that should probably be obvious from the description of strcspn(3)?

```
SYNOPSIS
     #include <string.h>

     ...
     size_t strcspn(const char *s, const char *reject);

DESCRIPTION
     ...

     The  strcspn()  function calculates the length of the initial segment of s
     which consists entirely of bytes not in reject.
```

To be super-precise, it returns a pointer to the first character in reject, or to the ending '\0' if none is found.

So, it's basically a strchrnul(3) that can search for several delimiters.

Reply:
Maybe?:

```
// Similar to strcspn(3), but return a pointer.
// Similar to strchrnul(3), but search for several delimiters.

Reply:
I've done this:

```
$ git range-diff master gh/string string 
1:  eeff3e63 ! 1:  d7cd8b88 lib/string/strchr/: stp[rc]spn(), strnul(): Add macros and functions
    @@ lib/string/strchr/stpcspn.h (new)
     +
     +
     +// Similar to strcspn(3), but return a pointer.
    ++// Similar to strchrnul(3), but search for several delimiters.
    ++// Similar to strpbrk(3), but return 's + strlen(s)' if not found.
     +#define stpcspn(s, reject)                                                    \
     +({                                                                            \
     +  __auto_type  s_ = s;                                                  \
    @@ lib/string/strchr/stprspn.h (new)
     +#include "string/strchr/strnul.h"
     +
     +
    -+// Similar to strspn(3), but start from the end, and return a pointer.
    ++// Similar to strspn(3),
    ++//  but start from the end, and return a pointer.
    ++// Similar to strrchr(3),
    ++//  but search for several delimiters, and return 's' if not found.
    ++// Identical to Solaris strrspn(3GEN),
    ++//  but I didn't like their choice of a name.
     +ATTR_STRING(2)
     +inline char *stprspn(char *restrict s, const char *restrict accept);
     +
    @@ lib/string/strchr/stpspn.h (new)
     +
     +
     +// Similar to strspn(3), but return a pointer.
    ++// Similar to strchrnul(3), but search for all bytes not in 'accept'.
     +#define stpspn(s, accept)                                                     \
     +({                                                                            \
     +  __auto_type  s_ = s;                                                  \
2:  460cae37 = 2:  36a95b16 lib/getdef.c: def_load(): Use stp[c]spn() instead of their patterns
3:  33f5eed8 = 3:  f9c7504f src/login_nopam.c: login_access(): Simplify, calling strchr(3)
4:  0866dd74 = 4:  3294c812 lib/sssd.c: Style fixes
5:  f5930587 = 5:  025353a6 lib/, src/: Use stprspn() instead of its pattern
6:  896d8843 = 6:  1963586f contrib, lib/, src/, tests/: Use stpcpy(3) instead of its pattern
7:  d6bfd265 = 7:  c39d3337 contrib/adduser.c: main(): Use strcpy/cat(3) instead of their pattern
```

Reply:
In any case, I expect the implementation to tell more than the description (and they're in the same place, so someone reading the description will necessarily read the implementation):

```c
$ grepc stpspn .
./lib/string/strchr/stpspn.h:#define stpspn(s, accept)                                                     \
({                                                                            \
	__auto_type  s_ = s;                                                  \
                                                                              \
	s_ + strspn(s_, accept);                                              \
})
```

Which looks like strspn(3) returns a length, and then we add it to the original pointer.  Now the job is to read the manual page for what strspn(3) is.